### PR TITLE
Added test for wall clock time change on the client

### PR DESCRIPTION
**Pull request by rmstar**

### CODE REVIEW

**Comment by rmstar** [comment]
Known issues: #17335, #17913

File: test/cpp/end2end/time_change_test.cc [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+}  // namespace
+
+}  // namespace testing
+}  // namespace grpc
+
+int main(int argc, char** argv) {
+  std::string me = argv[0];
+  auto lslash = me.rfind('/');
```
Comment (muxi) [inline test/cpp/end2end/time_change_test.cc]: I kind of feel `lslash` should be `rslash` here but I might not understood this correctly :)

Comment (muxi) [reply test/cpp/end2end/time_change_test.cc]: A comment here may be helpful to better understand the if-else block below

File: test/cpp/end2end/time_change_test.cc [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+  context.AddMetadata(kServerResponseStreamsToSend, "1");
+
+  auto channel = GetChannel();
+  GPR_ASSERT(channel);
+  EXPECT_TRUE(
+      channel->WaitForConnected(grpc_timeout_milliseconds_to_deadline(5000)));
+  auto stub = CreateStub();
+
+  // time jumps forward by 20.123 seconds.
+  set_now_offset(20123);
```
Comment (muxi) [inline test/cpp/end2end/time_change_test.cc]: Suggest make this a constant and define it somewhere else. Same for similar places below.

File: test/cpp/end2end/time_change_test.cc [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+  std::shared_ptr<Channel> GetChannel() { return channel_; }
+
+  std::unique_ptr<SubProcess> server_;
+  std::shared_ptr<Channel> channel_;
+  std::unique_ptr<grpc::testing::EchoTestService::Stub> stub_;
+};
+
+// Wall-clock time jumps forward on client before bidi stream is created
+TEST_F(TimeChangeTest, TimeJumpForwardBeforeStreamCreated) {
```
Comment (muxi) [inline test/cpp/end2end/time_change_test.cc]: Is it possible to produce multiple shifts in a test? Just more extreme cases that worth being tested and also covers the one-shift case.

File: test/cpp/end2end/time_change_test.cc (hunk 1 above)
Comment (rmstar) [reply test/cpp/end2end/time_change_test.cc]: I think `lslash` = 'last slash'. I got that piece of code from `test/cpp/end2end/server_crash_test.cc`. I have added a comment to explain what's going on here.

File: test/cpp/end2end/time_change_test.cc (hunk 2 above)
Comment (rmstar) [reply test/cpp/end2end/time_change_test.cc]: Done.

File: test/cpp/end2end/time_change_test.cc (hunk 3 above)
Comment (rmstar) [reply test/cpp/end2end/time_change_test.cc]: Added a new test case `TimeJumpForwardAndBackDuringCall` with multiple shifts during a call.

File: test/cpp/end2end/time_change_test.cc (hunk 1 above)
Comment (muxi) [reply test/cpp/end2end/time_change_test.cc]: Ah ok. I thought it was "left slash"

File: src/core/lib/gpr/time.cc [hunk 4]
```diff
@@ -135,6 +135,10 @@ gpr_timespec gpr_time_add(gpr_timespec a, gpr_timespec b) {
   gpr_timespec sum;
   int64_t inc = 0;
   GPR_ASSERT(b.clock_type == GPR_TIMESPAN);
+  // tv_nsec in a timespan is always +ve. -ve timespan is represented as (-ve
+  // tv_sec, +ve tv_nsec). For example, timespan = -2.5 seconds is represented
+  // as {-3, 5e8, GPR_TIMESPAN}
+  GPR_ASSERT(b.tv_nsec >= 0);
```
Comment (mehrdada) [inline src/core/lib/gpr/time.cc]: Are we confident this assertion is always true? We are seeing segfaults because of this.

**Pull request merged**